### PR TITLE
Serialize custom augmentations losslessly in checkpoints

### LINKED ISSUE #25766: RepresenterError and checkpoint corruption when using custom Albumentations transforms via augmentations kwarg
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Search before asking
- [x] I have searched the YOLO issues and discussions and found no similar bug report.

### Bug Report

When passing a list of custom Albumentations transforms (e.g., `A.MotionBlur`, `A.ImageCompression`) via the `augmentations` kwarg in `model.train()`, the training pipeline crashes before Epoch 1. 

The issue occurs because the trainer attempts to serialize the complex Python objects (Albumentations transforms) into the `args.yaml` file, causing a `yaml.representer.RepresenterError`. Furthermore, if this serialization step is bypassed, the transforms are pickled directly into `best.pt`, which corrupts the checkpoint and makes it unloadable on any host machine that does not have `albumentations` installed.

Currently, the only way to bypass this architecture flaw is to use an `on_pretrain_routine_start` callback to inject the transforms and manually remove them before each checkpoint save.

### Minimum Reproducible Example (MRE)

```python
from ultralytics import YOLO
import albumentations as A

# 1. Load a pretrained model
model = YOLO("yolov8n.pt") 

# 2. Define custom Albumentations transforms
custom_augs = [
    A.MotionBlur(p=0.15),
    A.ImageCompression(quality_lower=75, p=0.2)
]

# 3. Train the model (This will crash during YAML serialization)
model.train(
    data="coco8.yaml",
    epochs=3,
    augmentations=custom_augs
)
```

### Environment

- Ultralytics YOLO version: 8.4.116
- Python version: 3.11.15
- Albumentations version: 2.0.8
- PyTorch version: 2.6.0+cu124
- OS: Windows (RTX 4050 Laptop GPU, 6.4 GB VRAM)

### Minimal Reproducible Example

```
from ultralytics import YOLO
import albumentations as A

# 1. Load a pretrained model
model = YOLO("yolov8n.pt") 

# 2. Define custom Albumentations transforms
custom_augs = [
    A.MotionBlur(p=0.15),
    A.ImageCompression(quality_lower=75, p=0.2)
]

# 3. Train the model (This will crash during YAML serialization)
model.train(
    data="coco8.yaml",
    epochs=3,
    augmentations=custom_augs
)
```

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Rahulbiradar9**
### Summary
Custom Albumentations transforms passed via `augmentations=[...]` were pickled as live objects into `last.pt`/`best.pt` (`train_args` and `model.args`), which breaks restricted checkpoint loading and portability. This PR serializes them as `repr()` strings everywhere the run args are persisted, and makes resume handle those strings safely.

### Key Changes
- `BaseTrainer.serialized_args()` returns the run args with the `augmentations` list replaced by transform reprs; used for `args.yaml`, checkpoint `train_args`, and `ema.args` so no transform objects are pickled.
- `check_resume` clears serialized reprs (`self.args.augmentations = None`) when resuming without an `augmentations` override, logs the original transforms, and tells the user to pass the original transform objects again. An explicit `augmentations=[]` is preserved.

`augmentations` is a flat `list` of transforms, matching what `Albumentations(transforms=...)` consumes, so the one-level conversion is the complete contract.

Fixes #25766

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Rahulbiradar9, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed all three changed files and the related serialization, checkpoint loading, resume, and augmentation paths. The YAML and checkpoint sanitization fixes address the reported portability issue, but checkpoint reprs currently become active augmentation values during resume and can silently disable Albumentations. Resolve that resume path before merging.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/engine/trainer.py:751` These repr strings are fed back into the training configuration when resuming: `check_resume()` loads `train_args` and, if the user does not provide `augmentations`, leaves this list in `self.args`. `v8_transforms()` then passes the strings to `Albumentations`; construction fails and the class falls back to `transform=None`, so the resumed run silently disables Albumentations instead of using valid transforms. Keep the reprs only for serialized metadata and clear `self.args.augmentations` when…
</details>

**Comment by Rahulbiradar9** [comment]
# Albumentations Custom Transforms Serialization Fix
a complete summary of the bug, the root cause analysis, the implemented fixes, the formatting steps, and the testing verification.

## Root Cause Analysis
1. **`yaml.representer.RepresenterError`**: The Ultralytics training pipeline automatically serializes the training arguments to `args.yaml` via `YAML.save`. Since the `augmentations` parameter is passed as a list of custom Albumentations transforms (complex Python objects), PyYAML's safe dumper failed on them because it was only checking the top-level collection types.
2. **Checkpoint Corruption**: During training, the trainer serializes the model and state using `torch.save`, storing `vars(self.args)` as `train_args` in the checkpoint dictionary. Additionally, the training hyperparameters are attached directly to the model/EMA instance as `self.model.args` and `ema.args`. When custom transform objects were pickled, they made the checkpoint file dependent on local class definitions. This resulted in `AttributeError` or `ModuleNotFoundError` when loading the checkpoint in other environments.

---

## Implemented Fixes

### 1. Recursive YAML Serialization Sanitizer
* **File Modified**: [__init__.py](file:///D:/projcets/OPEN-SOURCE/ultralytics/ultralytics/utils/__init__.py)
* **Changes**:
  - Implemented a recursive helper function `sanitize()` inside `YAML.save` to walk dictionaries, lists, tuples, and sets, converting any non-serializable objects (like custom transforms) to their string representations.
  - Implemented the same recursive sanitization inside `YAML.print` for safe pretty-printing of dictionaries containing complex objects.

### 2. Checkpoint Serialization Sanitizer
* **File Modified**: [trainer.py](file:///D:/projcets/OPEN-SOURCE/ultralytics/ultralytics/engine/trainer.py)
* **Changes**:
  - Updated `save_model` to copy and serialize custom transform objects into repr strings in the `train_args` dictionary before calling `torch.save`.
  - Updated `save_model` to copy and sanitize `ema.args` (both namespace and dict structures) so that the serialized model checkpoint no longer contains picklable references to custom Python transform classes.

---

## Verification & Testing

### 1. Test Case Implementation
* **File Modified**: [test_python.py](file:///D:/projcets/OPEN-SOURCE/ultralytics/tests/test_python.py)
* **Changes**:
  - Added a test case `test_custom_albumentations` that:
    1. Instantiates a mock custom transform.
    2. Trains a tiny model (`imgsz=32`, `epochs=1`, `batch=2`) with `augmentations=[custom_transform]`.
    3. Verifies that `args.yaml` is written with string representations.
    4. Verifies that the saved checkpoint `last.pt` can be successfully loaded back using `YOLO(ckpt_path)` without corruption.

### 2. Test Execution
```bash
pytest tests/test_python.py -k test_custom_albumentations
```
* **Result**: **Passed** in `7.56s` with zero errors.

---

## Checklist
- [x] Tested with `pytest`
- [x] Code formatted with `ruff format .`
- [x] Linted with `ruff check .`

**Comment by pderrenger** [comment]
Please keep this fix scoped to `augmentations`; globally stringifying every unsupported value in `YAML.save()` and `YAML.print()` could silently mask invalid configuration data across the project. Extract a single serialization helper, apply it only to `train_args` and `ema.args`, and add focused tests for nested augmentation containers and checkpoint resume behavior.

**Comment by Rahulbiradar9** [comment]
@pderrenger 
let me see
i will tryout those changes

**Comment by Rahulbiradar9** [comment]
ihave updated the PR to address your feedback:

1. **Reverted global YAML changes**: Reverted the recursive serialization modifications in `YAML.save()` and `YAML.print()` to prevent silently masking invalid configuration data elsewhere.
2. **Extracted scoped serialization helper**: Added a single helper `serialize_augmentations` in `ultralytics/utils/__init__.py` to recursively serialize custom or non-serializable objects under the `augmentations` key/attribute while keeping python container structures (dicts, lists, sets, tuples) intact.
3. **Scoped application**: Applied the helper only to `train_args` and `ema.args` during checkpoint saving, and to `args_dict` when writing `args.yaml`.
4. **Added focused tests**: Added a unit test (`test_serialize_augmentations_nested`) specifically verifying serialization of nested containers, and kept the end-to-end checkpoint resume tests active.

All tests are passing locally. Let me know if this looks good!

**Comment by pderrenger** [comment]
This now matches the requested scope, including clearing serialized augmentation representations on resume unless valid transforms are overridden, and CI is passing. Please update the PR description to reflect the current scoped implementation rather than the reverted global `YAML` changes, then it is ready for final review.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 1, the resume path now clears serialized augmentation values when no override is supplied, addressing the prior finding. Two edge cases remain: empty augmentation lists are changed into default augmentations on resume, and the recovery warning presents non-restorable repr strings as a valid override example.

<details><summary>💬 Posted 2 inline comments</summary>

- 💡 **MEDIUM** `ultralytics/engine/trainer.py:1001` An explicit `augmentations=[]` is a valid way to disable Albumentations, but this condition treats every non-`None` value as unrecoverable custom transforms. Resuming without an override therefore changes `[]` to `None`, causing `Albumentations` to install its default transforms and silently changing the augmentation behavior. Only clear non-empty serialized augmentations here.
- 📝 **LOW** `ultralytics/engine/trainer.py:1009` The warning's example interpolates the serialized `repr` strings from `ckpt_args['augmentations']`, but those strings are not callable Albumentations transforms. Copying the displayed command causes Albumentations construction to fail and its exception handler disables the transform pipeline instead of restoring the original augmentations. The message should instruct users to recreate and pass the original transform objects, rather than presenting the serialized values as a usable override.
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 2, the empty-list resume case is addressed: `augmentations=[]` is preserved. The remaining issue is the recovery warning, which still presents non-restorable repr strings as a valid override and can lead users to silently disable Albumentations. No other genuine regressions found in the reviewed diff.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the trainer serialization and resume delta. The earlier empty-list resume regression is addressed; the remaining issue is the recovery warning, which still presents non-restorable repr strings as a valid override and can lead users to silently disable Albumentations.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 5

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since the previous review, the resume warning now correctly tells users to recreate the original transform objects, and the empty-list behavior remains fixed. The tracker and dense-result changes show no additional concrete regressions. One serialization gap remains: the current helper is not recursive and can corrupt nested augmentation configuration metadata.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/engine/trainer.py:712` This is only a one-level conversion, not the recursive sanitizer described by the PR. Every top-level element is converted with `repr()`, so nested YAML-compatible containers lose their structure and a non-list value such as a tuple or dict is iterated into an unrelated list of strings. Consequently `args.yaml`, `train_args`, and `ema.args` cannot faithfully represent nested augmentation configurations. Apply a recursive sanitizer to the `augmentations` value while preserving compatible contain…
</details>

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -709,6 +705,13 @@ def _model_train(self):
             if any(filter(lambda f: f in n, self.freeze_layer_names)) and isinstance(m, nn.BatchNorm2d):
                 m.eval()
 
+    def serialized_args(self) -> dict:
+        """Return a copy of the training args with custom augmentation transforms replaced by their reprs."""
+        args = vars(self.args).copy()
+        if args.get("augmentations") is not None:
+            args["augmentations"] = [repr(t) for t in args["augmentations"]]
```
Comment (glenn-jocher) [inline ultralytics/engine/trainer.py]: `augmentations` is a flat list of Albumentations transforms, exactly what `Albumentations(transforms=...)` in `ultralytics/data/augment.py` consumes, so a one-level `repr()` conversion is the complete contract. The recursive sanitizer was a description artifact from an earlier revision that was removed; the PR description is now updated to match the code.

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 6

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 5, the resume warning and explicit empty-list handling remain fixed. The one-level augmentation serialization matches the stated flat-list contract, and the dense-result, OCR, TrackTrack, YOLOE, test, and documentation changes show no additional concrete regressions. LGTM.
